src/realm/object-store/sync/sync_session.cpp

```diff
@@ -1236,12 +1257,79 @@ void SyncSession::update_configuration(SyncConfig new_config)
         REALM_ASSERT(m_state == State::Inactive || m_state == State::Paused);
         REALM_ASSERT(!m_session);
         REALM_ASSERT(m_config.sync_config->user == new_config.user);
+        // Since this is used for testing purposes only, just update the current sync_config
         m_config.sync_config = std::make_shared<SyncConfig>(std::move(new_config));
         break;
     }
     revive_if_needed();
 }
 
+void SyncSession::update_sync_config_after_migration()
+{
+    // Migration state changed - Update the configuration to
+    // match the new sync mode.
+    bool flx_sync_requested;
+    {
+        util::CheckedLockGuard cfg_lock(m_config_mutex);
+        m_config.sync_config = m_migration_store->convert_sync_config(m_original_sync_config);
+        flx_sync_requested = m_config.sync_config->flx_sync_requested;
+    }
+
+    util::CheckedLockGuard cfg_lock(m_state_mutex);
+    m_needs_subscription_store_updated.emplace(flx_sync_requested);
+}
+
+void SyncSession::update_subscription_store(bool flx_sync_requested)
+{
+    // The session should be closed before updating the FLX subscription store
+    REALM_ASSERT(!m_session);
+
+    // If the subscription store exists and switching to PBS, then clear the store
+    auto& history = static_cast<sync::ClientReplication&>(*m_db->get_replication());
+    if (!flx_sync_requested) {
+        if (m_flx_subscription_store) {
+            history.set_write_validator_factory(nullptr);
```

Comment:
So - thinking about this a little more - the write validator factory here just updates a UniqueFunction in a non-thread-safe way. We always got away with this because m_flx_subscription_store and the history validator were only ever updated in the constructor of SyncSession. Is this going to be unsafe now?

Reply:
Right, but you can have a Realm open with the sync session inactive tho. My point is that regardless of whats going on with the SyncSession, don't we need to block write transactions that might try to use the write validator?

Reply:
Right. We can easily protect `set_write_validator_factory` and `make_write_validator` with a lock, but the problem is the user can still commit changes. I think we have the same problem below where we create the subscription store. We can probably work around it if we set the validator from a write transaction.

Reply:
I tried adding realm transaction locking around the `set_write_validator_factory` calls, but got a TSAN error when the subscription store was being constructed, so I removed it (for now) - we'll need to revisit this in our next PRs.